### PR TITLE
Fix #2738 and #3067: Mastery calculations and tests

### LINKED ISSUE #2738: Add tests for score and mastery calculations
Need to write tests for [score and mastery calculation feature](https://docs.google.com/document/d/1ISuqvKs6b0xdE0-aEIHXYG7oDbCmDBcgUia0TiR9GpQ/edit?usp=sharing) after the calculation portion has been implemented. Tests should be checking the correctness of the final score and mastery degrees, since those are the only things relevant to the UI.

### LINKED ISSUE #3067: Implement Mastery Calculations for Practice Session
Implement mastery calculations in QuestionAssessmentCalculation, blocked by [score calculation](https://github.com/oppia/oppia-android/pull/3040) PR. This is part of the feature outlined in this [design document](https://docs.google.com/document/d/1ISuqvKs6b0xdE0-aEIHXYG7oDbCmDBcgUia0TiR9GpQ/edit?usp=sharing).

### PR DESCRIPTION
**Pull request by TheRealJessicaLi**
## Explanation
Added mastery calculations (with tests) for practice sessions. This fixes #3067 and #2738, but cannot be merged until PR #3040 (score calculations) is.

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Comment by BenHenning** [comment]
Sorry, been really busy with non-work items this week. Will need to take a look later in the week.

**Comment by BenHenning** [comment]
FYI that I'll be unavailable for reviews for the next ~18 days, so I probably won't be able to look at this until later. Please don't block on my review unless needed.

**Comment by vinitamurthi** [comment]
Hey @TheRealJessicaLi , assign me back once the bazel build passes and I will take a look!

File: domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt [hunk 1]
```diff
@@ -1286,6 +1286,214 @@ class QuestionAssessmentProgressControllerTest {
...
   }
 
+  @Test
+  fun solutionViewedForAllQuestions_returnMaxMasteryLossPerQuestion() {
+    setUpTestApplicationWithSeed(questionSeed = 0)
+    subscribeToCurrentQuestionToAllowSessionToLoad()
+    // This will generate question 1 (skill 0), question 2 (skill 0), and question 3 (skill 1)
+    startTrainingSession(TEST_SKILL_ID_LIST_01)
+
+    // Question 1
```
Comment (vinitamurthi) [inline domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt:1308]: Can you add a comment below that you are submitting a wrong answer here too?

File: domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt [hunk 2]
```diff
@@ -1286,6 +1286,214 @@ class QuestionAssessmentProgressControllerTest {
...
+    submitCorrectAnswerForQuestion1()
+
+    // Question 2
+    // submit question 2 wrong answer
+    submitIncorrectAnswerForQuestion2(4.0)
+    viewSolutionForQuestion2()
+    submitCorrectAnswerForQuestion2()
+
+    // Question 3
+    submitIncorrectAnswerForQuestion3("3/4")
```
Comment (vinitamurthi) [inline domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt:1322]: Same comment here

File: domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt [hunk 3]
```diff
@@ -1286,6 +1286,214 @@ class QuestionAssessmentProgressControllerTest {
...
+
+  @Test
+  fun hintsAndSolutionsViewedWithWrongAnswers_noMisconceptions_returnDifferingMasteryDegrees() {
+    setUpTestApplicationWithSeed(questionSeed = 0)
+    subscribeToCurrentQuestionToAllowSessionToLoad()
+    // This will generate question 1 (skill 0), question 2 (skill 0), and question 3 (skill 1)
+    startTrainingSession(TEST_SKILL_ID_LIST_01)
+
+    // Question 1
+    submitIncorrectAnswerForQuestion1(2)
```
Comment (vinitamurthi) [inline domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt:1371]: Add comments here too! Here and below

**Review by vinitamurthi — APPROVED** [review]
LGTM thanks! Just a few comments that need to be added but otherwise looks good

File: domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentCalculation.kt [hunk 4]
```diff
@@ -59,6 +65,36 @@ internal class QuestionAssessmentCalculation private constructor(
...
+      // Covert to Map<String, List<QuestionSessionMetrics>> for each linked skill ID.
+      { (skillId, _) -> skillId },
+      valueTransform = { (_, questionMetric) -> questionMetric }
+    ).mapValues { (skillId, questionMetrics) ->
+      // Compute the mastery change per question, converting type to Map<String, List<Int>>.
+      questionMetrics.map { it.computeMasteryChangeForSkill(skillId) }
+    }.mapValues { (_, masteryChanges) ->
+      // Convert to Map<String, Double> for each linked skill ID.
+      masteryChanges.sum().toDouble() / internalMasteryMultiplyFactor
+    }
```
Comment (rt4914) [inline domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentCalculation.kt:84]: Nice work

**Review by rt4914 — APPROVED** [review]
LGTM. Thanks.

File: domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt (hunk 1 above)
Comment (TheRealJessicaLi) [reply domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt:1308]: Done!

File: domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt (hunk 2 above)
Comment (TheRealJessicaLi) [reply domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt:1322]: Done

File: domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt (hunk 3 above)
Comment (TheRealJessicaLi) [reply domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt:1371]: Done

File: domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentCalculation.kt (hunk 4 above)
Comment (TheRealJessicaLi) [reply domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentCalculation.kt:84]: Thanks!

**Comment by TheRealJessicaLi** [comment]
@vinitamurthi I've updated my PR and all the tests are passing now, can you help me merge this in?

**Comment by vinitamurthi** [comment]
Thanks !Merging now

**Pull request merged**